### ISSUE TITLE
Poor results and remove pretrained classes from training

### ISSUE THREAD

**Issue by pauldeveaux** [ultralytics/ultralytics#13534]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Hello,

I am new to machine learning and YOLO, so I apologize if I'm not very accurate. I am trying to create an AI that can detect trash in images. For this, I have a dataset with a training set of 7217 images, a validation set of 525 images, and a test set of 227 images. I have only one class for trash

![image](https://github.com/ultralytics/ultralytics/assets/81611165/dc2a8cb2-8e9b-4ec3-b016-92d50584e3f5)

I am fine-tuning the YOLOv8n model, but I am getting poor results.

![image](https://github.com/ultralytics/ultralytics/assets/81611165/324bf630-baf2-4c4b-8836-6f5a41ae5a3c)

When I view the predictions on training images in wandb, I see that the model keeps predicting classes from the pre-trained model that I don't want. I don't know how to remove these classes (I think this might be causing the poor results, but I could be wrong).

![image](https://github.com/ultralytics/ultralytics/assets/81611165/b190c8b7-b4b8-4516-8129-d1d235504854)

Here is my code :

```
model = YOLO("/content/yolov8n.pt")

# Is GPU available
if torch.cuda.is_available():
    print("GPU is available")
else:
    print("GPU is not available")


weights_path = "/content/runs/detect/train2/weights/last.pt"
# Start or resume model
if os.path.exists(weights_path) and not RESTART:
    print("Resuming training from saved weights...")
    model = YOLO(weights_path)
else:
    print("Starting new training...")
    model = YOLO("/content/yolov8n.pt")


def save_on_drive(trainer):
  print(trainer.__dir__())
  epoch = trainer.epoch
  save_dir = "/content/drive/MyDrive/Model"
  save_path = os.path.join(save_dir, f'yolo_model_{epoch+1}.pt')
  model.save(save_path)


# wandb logs
add_wandb_callback(model, enable_model_checkpointing=True)

model.add_callback("on_train_epoch_end", save_on_drive)


results = model.train(
          data=os.path.join(dataset.location, "/content/Trash-Detector-2/data.yaml"),
          epochs=150,
          imgsz=640,
          batch=-1,
          workers=16,
          patience=50,
          resume=os.path.exists(weights_path)
      )

# Validate the model
model.val()

# Finish the wandb run
wandb.finish()
```

and the data.yaml : 

```
names:
- trash
nc: 1
roboflow:
  license: CC BY 4.0
  project: trash-detector-u1syo
  url: https://universe.roboflow.com/test-rlywk/trash-detector-u1syo/dataset/2
  version: 2
  workspace: test-rlywk
test: /content/Trash-Detector-2/test/images
train: /content/Trash-Detector-2/train
val: /content/Trash-Detector-2/valid
```

Can anyone help me understand why the model is still predicting classes from the pre-trained model and how to fix this?

Thank you!

**Comment by RS-v620** [comment]
There is a parameter ```single_cls``` in the ```model.train``` function that focuses the model towards detection instead of  classification. You can set it to ```True```. This will cause model to only detect the trash in images. 
Keep in mind that this will remove distinction between any classes you have in the dataset.

**Comment by pauldeveaux** [comment]
Hello, 

Thank you for your response. I did as you said (add the single_cls=True as argument in model.train() ). I had better results at the end of the training. 

```
results = model.train(
          data=os.path.join(dataset.location, "/content/Trash-Detector-2/data.yaml"),
          epochs=150,
          imgsz=640,
          batch=-1,
          workers=16,
          patience=50,
          single_cls=True,
          resume=os.path.exists(weights_path)
      )
```
However, when I print model.names, the "trash" class does not appear in the output:

```
{0: 'person', 1: 'bicycle', 2: 'car', 3: 'motorcycle', 4: 'airplane', 5: 'bus', 6: 'train', 7: 'truck', 8: 'boat', 9: 'traffic light', 10: 'fire hydrant', 11: 'stop sign', 12: 'parking meter', 13: 'bench', 14: 'bird', 15: 'cat', 16: 'dog', 17: 'horse', 18: 'sheep', 19: 'cow', 20: 'elephant', 21: 'bear', 22: 'zebra', 23: 'giraffe', 24: 'backpack', 25: 'umbrella', 26: 'handbag', 27: 'tie', 28: 'suitcase', 29: 'frisbee', 30: 'skis', 31: 'snowboard', 32: 'sports ball', 33: 'kite', 34: 'baseball bat', 35: 'baseball glove', 36: 'skateboard', 37: 'surfboard', 38: 'tennis racket', 39: 'bottle', 40: 'wine glass', 41: 'cup', 42: 'fork', 43: 'knife', 44: 'spoon', 45: 'bowl', 46: 'banana', 47: 'apple', 48: 'sandwich', 49: 'orange', 50: 'broccoli', 51: 'carrot', 52: 'hot dog', 53: 'pizza', 54: 'donut', 55: 'cake', 56: 'chair', 57: 'couch', 58: 'potted plant', 59: 'bed', 60: 'dining table', 61: 'toilet', 62: 'tv', 63: 'laptop', 64: 'mouse', 65: 'remote', 66: 'keyboard', 67: 'cell phone', 68: 'microwave', 69: 'oven', 70: 'toaster', 71: 'sink', 72: 'refrigerator', 73: 'book', 74: 'clock', 75: 'vase', 76: 'scissors', 77: 'teddy bear', 78: 'hair drier', 79: 'toothbrush'}
```

Could you please help me understand why the class is missing and how I can ensure it is included in the model?

Thank you for your assistance!

**Comment by glenn-jocher** [comment]
@pauldeveaux hello,

Thank you for your prompt response and for providing additional details. It's great to hear that using `single_cls=True` has improved your results. However, I understand the confusion regarding the `model.names` output.

The issue you're encountering is likely due to the model still retaining the class names from the pre-trained weights. When you fine-tune a model with a custom dataset, especially with a single class, it's essential to ensure that the model's class names are updated accordingly.

Here are a few steps to address this:

1. **Ensure Data Configuration**: Double-check your `data.yaml` file to ensure it correctly specifies the single class. It looks like your `data.yaml` is correctly set up, but just to confirm:
    ```yaml
    names:
    - trash
    nc: 1
    train: /content/Trash-Detector-2/train
    val: /content/Trash-Detector-2/valid
    test: /content/Trash-Detector-2/test/images
    ```

2. **Load Model with Custom Data**: When loading the model for training, ensure it reads the custom data configuration. This can be done by specifying the `data` argument correctly:
    ```python
    results = model.train(
        data='/content/Trash-Detector-2/data.yaml',
        epochs=150,
        imgsz=640,
        batch=-1,
        workers=16,
        patience=50,
        single_cls=True,
        resume=os.path.exists(weights_path)
    )
    ```

3. **Update Class Names**: After training, you can manually update the `model.names` to reflect your custom class:
    ```python
    model.names = {0: 'trash'}
    ```

4. **Check Model Names After Training**: Ensure that the model's class names are updated after training:
    ```python
    print(model.names)  # Should output {0: 'trash'}
    ```

Here is a revised version of your training script with these considerations:

```python
model = YOLO("/content/yolov8n.pt")

# Check if GPU is available
if torch.cuda.is_available():
    print("GPU is available")
else:
    print("GPU is not available")

weights_path = "/content/runs/detect/train2/weights/last.pt"
# Start or resume model
if os.path.exists(weights_path) and not RESTART:
    print("Resuming training from saved weights...")
    model = YOLO(weights_path)
else:
    print("Starting new training...")
    model = YOLO("/content/yolov8n.pt")

def save_on_drive(trainer):
    print(trainer.__dir__())
    epoch = trainer.epoch
    save_dir = "/content/drive/MyDrive/Model"
    save_path = os.path.join(save_dir, f'yolo_model_{epoch+1}.pt')
    model.save(save_path)

# wandb logs
add_wandb_callback(model, enable_model_checkpointing=True)
model.add_callback("on_train_epoch_end", save_on_drive)

results = model.train(
    data='/content/Trash-Detector-2/data.yaml',
    epochs=150,
    imgsz=640,
    batch=-1,
    workers=16,
    patience=50,
    single_cls=True,
    resume=os.path.exists(weights_path)
)

# Update model names to reflect the custom class
model.names = {0: 'trash'}

# Validate the model
model.val()

# Finish the wandb run
wandb.finish()
```

By following these steps, you should be able to ensure that the model correctly recognizes and outputs the 'trash' class. If you continue to experience issues, please let us know, and we can further investigate.

Thank you for your patience and dedication to improving your model. Happy training! 🚀
